src/Features/Core/Portable/ExtractMethod/SelectionResult.cs

```diff
@@ -119,59 +119,55 @@ public TExecutableStatementSyntax GetLastStatement()
-
-            bool CreateAsyncMethodWorker()
-            {
-                var firstToken = GetFirstTokenInSelection();
-                var lastToken = GetLastTokenInSelection();
-                var syntaxFacts = SemanticDocument.GetRequiredLanguageService<ISyntaxFactsService>();
-
-                for (var currentToken = firstToken;
-                    currentToken.Span.End < lastToken.SpanStart;
-                    currentToken = currentToken.GetNextToken())
-                {
-                    // [|
-                    //     async () => await ....
-                    // |]
-                    //
-                    // for the case above, even if the selection contains "await", it doesn't belong to the enclosing block
-                    // which extract method is applied to
-                    if (syntaxFacts.IsAwaitKeyword(currentToken)
-                        && !UnderAnonymousOrLocalMethod(currentToken, firstToken, lastToken))
-                    {
-                        return true;
-                    }
-                }
-
-                return false;
-            }
-        }
-
-        public bool ShouldCallConfigureAwaitFalse()
+        /// <summary>
+        /// Checks all of the nodes within the user's selection to see if any of them satisfy the supplied <paramref
+        /// name="predicate"/>. Will not descend into local functions or lambdas.
+        /// </summary>
+        /// <param name="predicate"></param>
+        /// <returns></returns>
+        private bool CheckNodesInSelection(Func<ISyntaxFacts, SyntaxNode, bool> predicate)
         {
-            var syntaxFacts = SemanticDocument.GetRequiredLanguageService<ISyntaxFactsService>();
+            var firstToken = this.GetFirstTokenInSelection();
+            var lastToken = this.GetLastTokenInSelection();
+            var span = TextSpan.FromBounds(firstToken.SpanStart, lastToken.Span.End);
 
-            var firstToken = GetFirstTokenInSelection();
-            var lastToken = GetLastTokenInSelection();
+            using var _ = ArrayBuilder<SyntaxNode>.GetInstance(out var stack);
+            stack.Push(this.GetContainingScope());
 
-            var span = TextSpan.FromBounds(firstToken.SpanStart, lastToken.Span.End);
+            var syntaxFacts = this.SemanticDocument.GetRequiredLanguageService<ISyntaxFactsService>();
 
-            foreach (var node in SemanticDocument.Root.DescendantNodesAndSelf())
+            while (stack.TryPop(out var current))
             {
-                if (!node.Span.OverlapsWith(span))
+                // Don't dive into lambdas and local functions.  They reset the async/await context.
+                if (syntaxFacts.IsAnonymousOrLocalFunction(current))
                     continue;
 
-                if (IsConfigureAwaitFalse(node) && !UnderAnonymousOrLocalMethod(node.GetFirstToken(), firstToken, lastToken))
+                if (current.Span.OverlapsWith(span) && predicate(syntaxFacts, current))
```

Comment:
>current.Span.OverlapsWith(span) [](http://example.com/codeflow?start=20&length=31)

isn't this already checked before the node was pushed onto the stack?

Reply:
hrm... i could potentially remove it.  iw as being paranoid.  let me see about that in followup.

Reply:
ok.  this can be removed.  will do in followup pr.